### ISSUE TITLE
Vue lifecycle issues 

### ISSUE THREAD

**Issue by digitaldan** [openhab/openhab-webui#1390]
While building out an enhanced  video widget i noticed that every time a widget is navigated to the Vue `mounted` lifecycle hook is called.  My understanding is that this  means Vue is creating a new instance of this component into the DOM.  If you navigate 10 times to the same widget, you will see the `mounted` function called 10 times, so there no re-use of the widget.   

Unfortunately `unmounted` is never being called when a widget is routed away from, either in a full page transition or a popup.   This leads to issues:

1. Leak of Vue components in the Dom that are never reused or cleaned up
2. Components that have background services like timers or video streams run indefinitely 
3. Because of 1 and 2, my guess is the UI starts to progressively slow down the longer it is up and used without a browser reload.  

I have a hacky workaround where i have a timer that updates a value for Vue to render in my widget, then another timer that fires if the component  stops rendering , which happens when its no longer visible, then i can at least cleanup, but the widget remains somewhere hidden until a browser reload.   I can't seem to find any other event to trigger a cleanup on.

**Comment by digitaldan** [comment]
Just a revision to this.  This is not happening in all cases, it depends on how you navigate.  `beforeDestroy` does get called on component that were navigated to from within a page and then back out of using the `back` nav button , including closing pop-ups on the page.  

For example, on my main homepage, if i click on a button on the main page that navigates to another page, then click back, `beforeDestroy` gets called, likewise if i trigger a popup to a page, then close it, `beforeDestroy` gets called.  But, if i use the side menu to navigate between pages, then the currently showed component/page does not get notified, and `beforeDestroy` does not get called.  The mixed behavior was confusing me, as it seemed to work and then not. 

So i don't think this is nearly as bad as i originally thought :-) But still would like to know how to fix it .

**Comment by hubsif** [comment]
Hi @digitaldan,
Framework7 keeps one page in the background, i.e. when you navigate to another page the current page is just shifted to background for quicker back navigation. That's why Vue's `beforeDestroy` does not get called before you move two pages forward. Is this maybe what you're experiencing?

The mix of two Frameworks, F7 and Vue and the use of the F7 router is probably what creates the confusion here. For the [SIP client widget](https://github.com/hubsif/openhab-webui/tree/sipclient) I also had to catch the immediate event of navigating away from a widget/page and found the solution in using the [F7 `pageBeforeOut` event](https://github.com/hubsif/openhab-webui/blob/c5b8abe370858bddb646ab6b241f20c7237eef40/bundles/org.openhab.ui/web/src/components/widgets/standard/oh-sipclient-card.vue#L201) instead of using Vue's events.

**Comment by digitaldan** [comment]
Thank @hubsif !  

>Framework7 keeps one page in the background,

yes, this seems to be the case, but i could not find where that was documented.  Whats odd is that it does not do that when routing from within a  page, so not using the navigation in the side menu but rather from say widget actions.  I realize though in the end, the `destroy` method is probably not going to be reliable.  

>the [F7 pageBeforeOut event](https://github.com/hubsif/openhab-webui/blob/c5b8abe370858bddb646ab6b241f20c7237eef40/bundles/org.openhab.ui/web/src/components/widgets/standard/oh-sipclient-card.vue#L201) i

Ah,  i was struggling to figure out a way to get this from within a component , thanks for the solution,  i will give that a try!

>For the [SIP client widget](https://github.com/hubsif/openhab-webui/tree/sipclient) I

Well thats pretty funny, i also have a SIP client i wrote that uses SIP.JS ( which is a long ago fork of JSSIP) , i have been using this in a web widget as i have not ported it yet to the UI.   Right now i am mainly using a modified (by me)  version of Linphone on a few android tablets that has a button to open my driveway gate during calls.  This is all hooked up through a Freeswitch container on my openHAB box.   Curious how you have this setup to answer through the web UI.

**Comment by hubsif** [comment]
You can have a look [at the forum](https://community.openhab.org/t/poc-sip-client-in-mainui-for-door-intercom/133903), though that’s a little outdated already. I’m going to open a pull request with updated gifs soon.

**Comment by digitaldan** [comment]
So the `pageBeforeOut` event works nicely for cleaning up.   Unfortunately, when you do navigate back to the page, none of the lifecycle methods get fired, so I can't start my stream again, did you encounter that?

**Comment by digitaldan** [comment]
>You can have a look [at the forum](https://community.openhab.org/t/poc-sip-client-in-mainui-for-door-intercom/133903),

Nice, I'll definitely play around with it.   I was thinking there are a few use cases where it would be nice to have routes that could be triggered by an item event, so pop up a call box, video stream, alarm panel , etc....  Habpanel had something similar.  Some sort of persisted widgets would be nice as well, where they load when the app loads and can be routed to, shown, reused, etc..

**Comment by ghys** [comment]
Great answers by @hubsif!

> yes, this seems to be the case, but i could not find where that was documented.

The only place I know where there's a mention of it seem to be here: https://v5.framework7.io/docs/view#view-parameters
(the `preloadPreviousPage` parameter).

> So the pageBeforeOut event works nicely for cleaning up. Unfortunately, when you do navigate back to the page, none of the lifecycle methods get fired, so I can't start my stream again, did you encounter that?

Interesting challenge, here's how I would do it, probably needs more testing:

```js
  data () {
    return {
      ...
      currentPageEl: null
    }
  },
  methods: {
    ...
    onPageAfterIn (page) {
      if (page.el === this.currentPageEl && page.direction === 'backward') {
        console.log('restarting')
        // restart stream
        this.currentPageEl = null
        this.$f7.off('pageAfterIn', this.onPageAfterIn)
        this.$f7.once('pageBeforeOut', this.onPageBeforeOut)
      }
      // console.dir(page)
    },
    onPageBeforeOut (page) {
      if (!this.currentPageEl) {
        this.currentPageEl = page.el
        // console.dir(page)
        this.$f7.on('pageAfterIn', this.onPageAfterIn)
      }
    }
  },
  mounted () {
    this.$f7.once('pageBeforeOut', this.onPageBeforeOut)
  },
  beforeDestroy () {
    this.$f7.off('pageAfterIn', this.onPageAfterIn)
    this.$f7.off('pageBeforeOut', this.onPageBeforeOut)
  }

```
In a nutshell:
1. you register your `pageBeforeOut` (fired only once) hook when the widget is mounted, so that when you leave the page it is fired
2. in the `pageBeforeOut` hook you store the current page element (from the [page data](https://v5.framework7.io/docs/page#page-data)), and also register your hook for `pageAfterIn` (you can probably use `pageBeforeIn` at the risk of the back animation not being smooth) if there's no current page to be navigated back to
3. in the `onAfterIn` you first check that we're navigating back to the page, and if so, perform your restart logic, but also clear the page reference and register the `pageBeforeOut` hook again
4. make sure you cleanup event subscriptions when the widget is effectively destroyed

**Comment by digitaldan** [comment]
Thanks, i'll need to play with it some more, my initial attempt at using the example above has run into issues with properties and state in the widget being null, using nextTick in some cases helps, but in some cases not.  Navigation is not consistant either when using back vs clicking through pages from the sidebar menu.  I need to debug more before i'm making any sense here.

**Comment by ghys** [comment]
> clicking out to another page then going back directly to this page does not (using the side menu links)

I suppose you've tried removing the `&& page.direction === 'backward'` test in the code above? Maybe the `onPageAfterIn` function is called but does nothing?
When going back and forth from the menu it would not be true (but you would still have 2 pages in the DOM, excluding modals, in case the browser's back button or swipe-to-back gesture is used).
All things considered the "direction" doesn't matter at all, what matters is that the page was brought into view again.

**Comment by digitaldan** [comment]
I have tried a bunch of permutations , including remove the backwards check,  that actually may have worked at some point (i'm in that chaotic cycle of trial and error programming which is not pretty) , but then another problem that happens  is that if I click through 3 pages, then click back, the first page gets (re)mounted unexpectedly, for example given the following routes: 

page1 --> page 2 --> page3  (click `back`) -->  page 2 

when i click `back ` to page 2, `mount` gets called on page 1, even though we are on page 2 , its like f7 is trying to preload the previous page in the history, but since `mount` happens at component creation and there's no listeners, i'm not sure how to  tell that new widget not to stream?   This scenario seems very difficult to solve.   

Btw i appreciate the help, this is surprisingly baffling to me 😵‍💫

**Comment by ghys** [comment]
Yes I believe it's the exact purpose of that `preloadPreviousPage` parameter in https://v5.framework7.io/docs/view#view-parameters, which is set to true by default.

An example with settings pages - but user-defined pages follow the same principle:

![chrome_2022-05-17_14-25-29](https://user-images.githubusercontent.com/2004147/168814057-890d236c-35d5-4151-adc3-cdf5dbaf2068.gif)

Here I "go deep" in navigation; as I understand it, it means that when you go to page 3 (the rule page), the first one (the Settings page) is destroyed because Framework7 only keeps the immediate previous page in the DOM (there's a `keepAlive` flag to prevent this behavior for a particular page), however when I go back to page 2 (the rules list), and then go back again to the Settings page, there's still a working "going back" animation. How? That's because page 1 has been mounted again in the background after going from page 3 to page 2, to allow the transition to page 1 to work, should the user choose to go back again.

Obviously this is a problem when there are computationally heavy elements on a page; you don't want them to run unless the page is in the foreground - how to determine that is the question. This is a problem that I believe deserves a "there's the recommended best practice" answer so I'll get back to you :)

**Comment by digitaldan** [comment]
> This is a problem that I believe deserves a "there's the recommended best practice" answer so I'll get back to you :)

Thats music to my ears!

**Comment by digitaldan** [comment]
Hi @ghys any thoughts on a solution yet?

**Comment by ghys** [comment]
@digitaldan apologies for the delay, I hadn't forgotten, but you know how it is 🙈

I have come up with a minimal example of a widget that detects whether it is currently in the foreground or not, along with hooks when the transition happens, and I haven't been able to fault it (...yet, but you might).
The code is even more straightforward compared to my earlier suggestion.

Please check it out here: https://github.com/openhab/openhab-webui/compare/main...ghys:widget-detect-foreground

The bodies of `this.startForegroundActivity` and `this.stopForegroundActivity` are meant to be replaced with whatever needs to happen in your widget when it enters or leaves the foreground. In fact maintaining that `this.foreground` flag is not strictly necessary if you just need the transition hooks and don't need the actual state.

Also note that the "back" transition's duration can take an arbitrary time when the "swipe right to go back" gesture is activated. Here the `onAfterIn` page event is considered so the foreground activity only starts *after* the page has transitioned in, as illustrated below; if you wanted to start as soon as the page is even partly shown you would use `onBeforeIn` instead (but beware of transitions not going smooth).

![chrome_2022-06-03_20-02-14](https://user-images.githubusercontent.com/2004147/171920931-7f3d951e-7a97-45da-822c-777b443e55a6.gif)

**Comment by digitaldan** [comment]
@ghys brilliant !  yes this seems to work in all cases, and yes the foreground variable is needed in my case when clicking `back` as  when it preloads the  previous page in history (we discussed previously) `mount` gets called with all the props set to null, and then `watch` gets called later with the values, so you need to know if the widget is in the foreground and needs to restart due to a src change (which can happen if an item event changes the src, etc...). 

I'll update this PR and submit.  I'll also submit a second PR for the image widget to stop the optional refresh timer there. 

> I hadn't forgotten, but you know how it is 🙈

Thanks for taking the time to figure that out , i definitely know how it is 😄  and appreciate it!

**Comment by florian-h05** [comment]
@digitaldan Not sure if that’s the right place, but it seems like on iOS Safari, the mixin doesn’t correctly detect when the page is left. 
I’m using the SIP client component, and it should unconnect and do not use the microphone anymore. On Android, this is working fine (so I think the problem is not the SIP client), but on iOS the microphone is still accessed when I switch to another page on openHAB.
